The gem was incorrectly parsing the book of Jude

codemis opened the pull request:
## The Problem

I discovered this issue on the bible_api issue tracker on [Issue #25](https://github.com/seven1m/bible_api/issues/25).  When the gem attempts to parse **Jude 1** it only returns verse 1.  However, when you parse other 1 chapter books like **Philemon 1** or **2 John 1**, you receive all of the verses.  Here is an example demonstrating the issue:

```
require 'bundler'

Bundler.require
ref = BibleRef::Reference.new('John 1')

puts '-- John 1 --'
puts ref.ranges.inspect

ref = BibleRef::Reference.new('Philemon 1')

puts '--- Philemon 1 ---'
puts ref.ranges.inspect

ref = BibleRef::Reference.new('2 John 1')

puts '--- 2 John 1 ---'
puts ref.ranges.inspect

ref = BibleRef::Reference.new('Jude 1')

puts '--- Jude 1 ---'
puts ref.ranges.inspect
```

When I ran this code, here were my results:

```
-- John 1 --
[[{:book=>"JHN", :chapter=>1}, {:book=>"JHN", :chapter=>1}]]
--- Philemon 1 ---
[[{:book=>"PHM", :chapter=>1}, {:book=>"PHM", :chapter=>1}]]
--- 2 John 1 ---
[[{:book=>"2JN", :chapter=>1}, {:book=>"2JN", :chapter=>1}]]
--- Jude 1 ---
[[{:book=>"JUD", :chapter=>1, :verse=>1}, {:book=>"JUD", :chapter=>1, :verse=>1}]]
```

As you can see, Jude returns verse 1 in the ranges which is incorrect.

## The Fix

I removed some old code that was designed to fix how Jude was parsed, but it actually caused Jude to return the wrong response.  I was able to also fix one test that was being skipped due to this fix.  Now when I run my test script from above, here are my results:

```
-- John 1 --
[[{:book=>"JHN", :chapter=>1}, {:book=>"JHN", :chapter=>1}]]
--- Philemon 1 ---
[[{:book=>"PHM", :chapter=>1}, {:book=>"PHM", :chapter=>1}]]
--- 2 John 1 ---
[[{:book=>"2JN", :chapter=>1}, {:book=>"2JN", :chapter=>1}]]
--- Jude 1 ---
[[{:book=>"JUD", :chapter=>1}, {:book=>"JUD", :chapter=>1}]]
```

I hope this helps.  Let me know you have any questions.

seven1m commented:
I believe the original "fix" was trying to make it so people can reference Jude verses without a chapter, e.g. http://bible-api.com/jude+3

I think it should be possible to fix the code so that both "Jude 3" and "Jude 1:3" works as expected.

I know this creates a bit of ambiguity when someone requests Jude 1 -- do we return the full chapter or just the verse? I think we just have to return the single verse, as that is probably what most people want.

Would you care to rework your PR so that both of the following work?

- "Jude X" - return single verse from chapter 1
- Jude 1:X" - return single verse from chapter 1

codemis commented:
Thanks for your reply.  Sure, I can look into the fix.  Should I apply the same fix to all 1 chapter books?  That way we have consistency.

seven1m commented:
Yeah, I think that'd be great! Thanks! ❤️

codemis commented:
Ok, I was able to get it to work with either Jude 1:1-5 or Jude 1-5.  When the script initializes the Reference class, I check if it is a single chapter book, and if so I standardize the reference.  Basically, I add the chapter 1 to the reference.  So **Jude 1-5** becomes **Jude 1:1-5** before being fed to the parser.  This works for all single chapter books.  Thoughts?

seven1m commented on lib/bible_ref/languages/english.rb:
```diff
@@ -88,7 +95,7 @@ def books
           '2JN' => { match: /^2 ?jo?h?n/,        name: '2 John'                 },
           '3JN' => { match: /^3 ?jo?h?n/,        name: '3 John'                 },
-          'JUD' => { match: /^(jud$|jd$|jude$)/, name: 'Jude'                   },
+          'JUD' => { match: /^(jud|jd|jude)/,    name: 'Jude'                   },
```
I'm a bit nervous about this change, given it was just fixed in #7. Can you make sure that the existing behavior of differentiating between Judith and Jude is left intact. (If not, we may need a regression test to keep it from happening in the future.)

codemis replied:
Sure.  I will write a test to validate it. 🤓👍

seven1m commented on lib/bible_ref/languages/english.rb:
```diff
@@ -4,6 +4,13 @@
+      # Is it a single chapter book?
+      def has_single_chapter?(reference)
+          matches = [/^ob/, /^(jud|jd|jude)/, /^2 ?jo?h?n/, /^3 ?jo?h?n/, /^ph(i?l|m)/]
+          return Regexp.union(matches).match?(reference.downcase)
```
TIL `Regex.union` :heart_eyes:

codemis commented:
Ok.  I fixed the Regex statement, and added some code to not mark Judith as a single chapter book.  I also updated the README so developers know what to expect from single chapter books.

seven1m commented:
Thanks! I'll cut a release and deploy this to bible-api.com later today...

seven1m commented:
bible-api.com is now updated with this code! Thanks for your help!
